java/feathub-udf/flink-udf/src/main/java/com/alibaba/feathub/flink/udf/aggregation/UnboundedOverWindowAggFuncWithLimit.java

```diff
@@ -0,0 +1,94 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.alibaba.feathub.flink.udf.aggregation;
+
+import org.apache.flink.api.java.tuple.Tuple2;
+import org.apache.flink.table.types.DataType;
+import org.apache.flink.util.Preconditions;
+
+/**
+ * The UnboundedRangeOverWindowAggFuncWithLimit is used in unbounded range over window, where the
+ * input is sorted before adding to the accumulator so that we only need to keep up to limit number
+ * of values in the accumulator.
+ *
+ * <p>See {@link
+ * org.apache.flink.table.runtime.operators.over.RowTimeRangeUnboundedPrecedingFunction}.
+ */
+public class UnboundedOverWindowAggFuncWithLimit<IN_T, OUT_T, ACC_T>
+        extends RawDataAccumulatingAggFunc<IN_T, OUT_T> {
+
+    private final AggFunc<IN_T, OUT_T, ACC_T> innerAggFunc;
+    private final long limit;
+
+    public UnboundedOverWindowAggFuncWithLimit(
+            AggFunc<IN_T, OUT_T, ACC_T> innerAggFunc, long limit) {
+        this.innerAggFunc = innerAggFunc;
+        this.limit = limit;
+    }
+
+    @Override
+    public void add(RawDataAccumulator<IN_T> acc, IN_T value, long timestamp) {
+        Preconditions.checkState(
+                acc.rawDataList.isEmpty() || timestamp >= acc.rawDataList.getLast().f1,
+                "The records should be sorted in the unbounded event time over window.");
+        super.add(acc, value, timestamp);
+
+        while (acc.rawDataList.size() > limit) {
+            acc.rawDataList.removeFirst();
+        }
+    }
+
+    @Override
+    public void merge(RawDataAccumulator<IN_T> target, RawDataAccumulator<IN_T> source) {
+        super.merge(target, source);
+
+        while (target.rawDataList.size() > limit) {
+            target.rawDataList.removeFirst();
+        }
+    }
+
+    @Override
+    public OUT_T getResult(RawDataAccumulator<IN_T> accumulator) {
+        Preconditions.checkState(accumulator.rawDataList.size() <= limit);
+
+        ACC_T acc = innerAggFunc.createAccumulator();
+
+        for (Tuple2<IN_T, Long> data : accumulator.rawDataList) {
+            innerAggFunc.add(acc, data.f0, data.f1);
+        }
+
+        return innerAggFunc.getResult(acc);
+    }
+
+    @Override
+    public DataType getResultDatatype() {
+        return innerAggFunc.getResultDatatype();
+    }
+
+    @Override
+    public void retractAccumulator(
+            RawDataAccumulator<IN_T> target, RawDataAccumulator<IN_T> source) {
+        // Event time unbounded over window will not retract.
+        throw new UnsupportedOperationException();
```

Comment:
The `retract` and `restractAccumulator` must not be invoked. I want the failed fast behavior in case it happens. I updated the exception message to explicitly state that.